pkg/expansion/db.go

```diff
@@ -0,0 +1,291 @@
+		}
+	}
+	return confs
+}
+
+// handleAdd adds the template to the DB, returning true if a cycle was created.
+// The template is added even if a cycle was found.
+func (d *db) handleAdd(template *expansionunversioned.ExpansionTemplate) (bool, error) {
+	id := keyForTemplate(template)
+
+	// We should always remove the old template before handleAdd. If we
+	// didn't, that's a bug.
+	if _, exists := d.store[id]; exists {
+		panic(fmt.Errorf("tried to add template %q that already exists", id))
+	}
+
+	// Update storage
+	d.store[id] = &templateState{template: template.DeepCopy(), hasConflicts: false}
+
+	// Update generators
+	genGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+	if _, exists := d.generators[genGVK]; !exists {
+		d.generators[genGVK] = make(map[TemplateID]bool)
+	}
+	d.generators[genGVK][id] = true
+
+	// Update matchers
+	matches := applyToGVKs(template)
+	for _, m := range matches {
+		if _, exists := d.matchers[m]; !exists {
+			d.matchers[m] = make(map[TemplateID]bool)
+		}
+		d.matchers[m][id] = true
+	}
+
+	// Add vertex if DNE
+	if _, err := d.graph.Vertex(hashID(id)); err != nil {
+		if errors.Is(err, graph.ErrVertexNotFound) {
+			if err := d.graph.AddVertex(id); err != nil {
+				return false, fmt.Errorf("adding vertex to graph: %w", err)
+			}
+		} else {
+			return false, fmt.Errorf("unexpected error getting vertex for template %s: %w", id, err)
+		}
+	}
+
+	// Add edges
+	edges := d.edgesForTemplate(template)
+	cycle := false
+	for _, e := range edges {
+		from := hashID(e.x)
+		to := hashID(e.y)
+
+		createsCycle, err := graph.CreatesCycle(d.graph, from, to)
+		if err != nil {
+			return false, fmt.Errorf("checking cycle for template %s: %w", id, err)
+		}
+		cycle = createsCycle || cycle
+
+		if err = d.graph.AddEdge(from, to); err != nil {
+			return false, fmt.Errorf("adding edge for template %s: %w", id, err)
+		}
+	}
+
+	return cycle, nil
+}
+
+func (d *db) edgesForTemplate(template *expansionunversioned.ExpansionTemplate) []edge {
+	var edges []edge
+	id := keyForTemplate(template)
+	genGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+
+	// Add out-bound edges (from this template's generated GVK to other
+	// templates' matched GVKs)
+	for t := range d.matchers[genGVK] {
+		edges = append(edges, edge{id, t})
+	}
+
+	// Add in-bound edges (from other templates' generated GVK to this template's
+	// matched GVK)
+	for _, gen := range applyToGVKs(template) {
+		for t := range d.generators[gen] {
+			edges = append(edges, edge{t, id})
+		}
+	}
+
+	return edges
+}
+
+func (d *db) handleRemove(template *expansionunversioned.ExpansionTemplate) {
+	id := keyForTemplate(template)
+
+	// The template must exist. Existence checks should be done upstream.
+	if _, exists := d.store[id]; !exists {
+		panic(fmt.Errorf("called handleRemove for template %q, but template DNE in store", id))
+	}
+
+	// Update storage
+	delete(d.store, id)
+
+	// Update generators
+	genGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
```

Comment:
We should be using the internally cached version of this template in order to guarantee consistency

Reply:
yep, good thinking. Changed `handleRemove` to just take a `TemplateID` as an arg, use that to lookup the cached version, and then remove the cached version